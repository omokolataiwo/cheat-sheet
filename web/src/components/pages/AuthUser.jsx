import React, { Component } from 'react';
import { connect } from 'react-redux';
import Masonry from 'react-masonry-component';
import { fetchFavoriteCheats } from '../../actions/cheat';
import Cheat from '../container/Cheat';

export class AuthUser extends Component {
  state = {};

  componentDidMount() {
    const { fetchFavoriteCheats } = this.props;
    const isAuthorizedUser = localStorage.getItem('token');

    if (isAuthorizedUser) {
      fetchFavoriteCheats();
    }
  }

  componentDidUpdate() {
    const isAuthorizedUser = localStorage.getItem('token');
    const {
      history: { push }
    } = this.props;

    if (!isAuthorizedUser) {
      push('/home');
    }
  }

  renderFavoriteCheat = () => {
    const { cheat: { favorite } } = this.props;

    if (!favorite.length) {
      return (<div className="no-favorite-cheat">You don't have favorite cheat yet.</div>);
    }
    return (
      <Masonry key={new Date()} className="cheat-masonry-grid">
        {localStorage.getItem('token')
        && favorite.map(favorite => (
          <div className="category-section">
            <Cheat cheat={favorite} showActionBar={false} />
          </div>
        ))}
      </Masonry>
    );
  }

  render() {
    const {
      history: { push }
    } = this.props;

    return (
      <div className="container">
        <h4>Favorite Cheats</h4>
        <hr />
        {this.renderFavoriteCheat()}
      </div>
    );
  }
}

const mapStateToProps = ({ user, cheat }) => ({ user, cheat });
export default connect(
  mapStateToProps,
  { fetchFavoriteCheats }
)(AuthUser);
