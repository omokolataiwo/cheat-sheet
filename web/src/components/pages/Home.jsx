import React from 'react';
import { connect } from 'react-redux';
import propTypes from 'prop-types';
import validate from 'validate.js';
import Masonry from 'react-masonry-component';
import Category from '../container/Category';
import SearchBar from '../container/SearchBar';
import { fetchAllCategoriesWithCheats, addToFavorite, searchSheetCheat } from '../../actions/cheat';
import { signin, signup, resetUserState } from '../../actions/user';
import { SIGNIN_SUCCESSFUL, SIGNOUT_SUCCESSFULLY, SIGNUP_FAILED } from '../../actions/type';
import SideNavPage from '../container/SideNavPage';
import signinValidationConstraint from '../../validatorConstraint/signin';
import signupValidationConstraints from '../../validatorConstraint/signup';

class Home extends React.Component {
  static propTypes = {
    cheat: propTypes.shape({
      cheats: propTypes.arrayOf()
    }).isRequired,
    signin: propTypes.func.isRequired,
    user: propTypes.shape().isRequired,
    history: propTypes.shape({ push: propTypes.func.isRequired }).isRequired
  };

  state = {
    user: {
      firstName: '',
      lastName: '',
      username: '',
      password: ''
    },
    errors: {}
  };

  componentDidUpdate() {
    const {
      user,
      resetUserState,
      history: { push }
    } = this.props;
    if (user.type === SIGNIN_SUCCESSFUL || user.type === SIGNOUT_SUCCESSFULLY) {
      return push('/user');
    }
    if (user.type === SIGNUP_FAILED) {
      resetUserState();
      this.setState(prevState => ({
        errors: { ...prevState.errors, signup: { ...user.errors.signup } }
      }));
    }
  }

  componentWillUnmount() {
    const { searchSheetCheat } = this.props;
    searchSheetCheat('');
  }

  addToFavorite = (id) => {
    const { addToFavorite } = this.props;
    addToFavorite(id);
  };

  formFieldChanged = ({ target: { value, name } }) => {
    this.setState(prevState => ({
      user: { ...prevState.user, [name]: value }
    }));
  };

  signin = (event) => {
    event.preventDefault();
    const { user } = this.state;
    const { signin } = this.props;
    const validationError = validate(user, signinValidationConstraint);

    this.setState(prevState => ({
      errors: { ...prevState.errors, signin: { ...validationError } }
    }));

    if (validationError) {
      return;
    }
    signin(user);
  };

  signup = (event) => {
    event.preventDefault();
    const { user } = this.state;
    const { signup } = this.props;

    const validationError = validate(user, signupValidationConstraints);

    this.setState(prevState => ({
      errors: { ...prevState.errors, signup: { ...validationError } }
    }));

    if (validationError) {
      return;
    }

    signup(user);
  };

  onSearchSheetCheat = (event) => {
    event.preventDefault();
    const { value } = event.target;
    const { searchSheetCheat } = this.props;
    searchSheetCheat(value.trim());
  };

  render() {
    let {
      cheat: { categories, searchResult },
      user: userAction
    } = this.props;
    const { user, errors } = this.state;

    categories = categories || [];
    categories = searchResult || categories;

    return (
      <div className="container">
        <SideNavPage
          onFormFieldChange={this.formFieldChanged}
          user={user}
          errors={errors}
          onSignin={this.signin}
          onSignup={this.signup}
          userAction={userAction}
        />
        <SearchBar onSearchSheetCheat={this.onSearchSheetCheat} />
        <Masonry key={new Date()} className="cheat-masonry-grid">
          {categories.map(category => (
            <Category category={category} onAddToFavorite={this.addToFavorite} />
          ))}
        </Masonry>
      </div>
    );
  }
}

export default connect(
  ({ cheat, user }) => ({ cheat, user }),
  {
    fetchAllCategoriesWithCheats,
    addToFavorite,
    signin,
    signup,
    searchSheetCheat,
    resetUserState
  }
)(Home);
