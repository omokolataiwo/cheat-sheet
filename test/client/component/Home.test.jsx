import React from 'react';
import { shallow } from 'enzyme';
import { Home } from '../../../web/src/components/pages/Home';
import {
  INITIAL_USER_STATE,
  SIGNIN_SUCCESSFUL,
  SIGNUP_FAILED
} from '../../../web/src/actions/type';

const props = {
  history: { push: jest.fn(() => {}) },
  addToFavorite: jest.fn(() => {}),
  signin: jest.fn(() => {}),
  signup: jest.fn(() => {}),
  resetUserState: jest.fn(() => { }),
  searchSheetCheat: jest.fn(() => {}),
  user: { type: INITIAL_USER_STATE },
  cheat: {
    errors: {},
    categories: [
      { title: 'Installation', cheats: [{ description: 'Initialize a repo', line: 'git init' }] }
    ]
  }
};
const wrapper = shallow(<Home {...props} />);

describe('Home Component', () => {
  it('should render component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should add cheat id as favorite', () => {
    const instance = wrapper.instance();
    instance.addToFavorite(12);
    expect(props.addToFavorite).toHaveBeenCalledWith(12);
  });

  it('should update state on field change', () => {
    const instance = wrapper.instance();
    instance.formFieldChanged({ target: { value: 'Taiwo', name: 'firstName' } });
    expect(wrapper.state().user.firstName).toEqual('Taiwo');
  });

  it('should signin user', () => {
    wrapper.setState({
      user: {
        firstName: 'Taiwo',
        lastName: 'Kolawole',
        username: 'omokolataiwo@gmail.com',
        password: ''
      }
    });
    const instance = wrapper.instance();
    instance.signin({ preventDefault: () => {} });
    expect(wrapper.state().errors).toEqual({ signin: { password: ['Password is too short'] } });
    wrapper.setState(prevState => ({ user: { ...prevState.user, password: '3232djsd' } }));
    instance.signin({ preventDefault: () => {} });
    expect(props.signin).toHaveBeenCalledWith(wrapper.state().user);
  });

  it('should sign up user', () => {
    wrapper.setState({
      user: {
        firstName: 'Taiwo',
        lastName: 'Kolawole',
        username: 'omokolataiwo@gmail.com',
        password: ''
      }
    });
    const instance = wrapper.instance();
    instance.signup({ preventDefault: () => {} });
    expect(wrapper.state().errors).toEqual({
      signin: {},
      signup: { password: ['Password is too short'] }
    });
    wrapper.setState(prevState => ({ user: { ...prevState.user, password: '3232djsd' } }));
    instance.signup({ preventDefault: () => {} });
    expect(props.signin).toHaveBeenCalledWith(wrapper.state().user);
  });

  it('should search sheet cheat', () => {
    const instance = wrapper.instance();
    instance.onSearchSheetCheat({ preventDefault: () => {}, target: { value: 'repo' } });
    expect(props.searchSheetCheat).toHaveBeenCalledWith('repo');
  });

  it('should redirect when user signin or signout', () => {
    wrapper.setProps({ user: { type: SIGNIN_SUCCESSFUL } });
    expect(props.history.push).toHaveBeenCalledWith('/user');
  });

  it('should set error on signin state when signup failed', () => {
    props.resetUserState = jest.fn(() => {
      wrapper.setProps({ user: { type: INITIAL_USER_STATE } });
    });

    wrapper.setProps({
      ...props,
      user: { type: SIGNUP_FAILED, errors: { signup: { firstName: ['Invalid username or password'] } } }
    });
    expect(wrapper.state().errors.signup).toEqual({ firstName: ['Invalid username or password'] });
    wrapper.unmount();
  });
});
