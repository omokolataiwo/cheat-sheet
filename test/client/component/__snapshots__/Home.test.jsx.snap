// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should render component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    addToFavorite={[MockFunction]}
    cheat={
      Object {
        "categories": Array [
          Object {
            "cheats": Array [
              Object {
                "description": "Initialize a repo",
                "line": "git init",
              },
            ],
            "title": "Installation",
          },
        ],
        "errors": Object {},
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    resetUserState={[MockFunction]}
    searchSheetCheat={[MockFunction]}
    signin={[MockFunction]}
    signup={[MockFunction]}
    user={
      Object {
        "type": Symbol(INITIAL_USER_STATE),
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SideNavPage
          errors={Object {}}
          onFormFieldChange={[Function]}
          onSignin={[Function]}
          onSignup={[Function]}
          user={
            Object {
              "firstName": "",
              "lastName": "",
              "password": "",
              "username": "",
            }
          }
          userAction={
            Object {
              "type": Symbol(INITIAL_USER_STATE),
            }
          }
        />,
        <SearchBar
          onSearchSheetCheat={[Function]}
        />,
        <MasonryComponent
          className="cheat-masonry-grid"
          disableImagesLoaded={false}
          elementType="div"
          enableResizableChildren={false}
          imagesLoadedOptions={Object {}}
          onLayoutComplete={[Function]}
          onRemoveComplete={[Function]}
          options={Object {}}
          updateOnEachImageLoad={false}
        >
          <Category
            category={
              Object {
                "cheats": Array [
                  Object {
                    "description": "Initialize a repo",
                    "line": "git init",
                  },
                ],
                "title": "Installation",
              }
            }
            onAddToFavorite={[Function]}
          />
        </MasonryComponent>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "errors": Object {},
          "onFormFieldChange": [Function],
          "onSignin": [Function],
          "onSignup": [Function],
          "user": Object {
            "firstName": "",
            "lastName": "",
            "password": "",
            "username": "",
          },
          "userAction": Object {
            "type": Symbol(INITIAL_USER_STATE),
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onSearchSheetCheat": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Mon Jan 21 2019 13:38:26 GMT+0100 (West Africa Standard Time)",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Category
              category={
                Object {
                  "cheats": Array [
                    Object {
                      "description": "Initialize a repo",
                      "line": "git init",
                    },
                  ],
                  "title": "Installation",
                }
              }
              onAddToFavorite={[Function]}
            />,
          ],
          "className": "cheat-masonry-grid",
          "disableImagesLoaded": false,
          "elementType": "div",
          "enableResizableChildren": false,
          "imagesLoadedOptions": Object {},
          "onLayoutComplete": [Function],
          "onRemoveComplete": [Function],
          "options": Object {},
          "updateOnEachImageLoad": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "category": Object {
                "cheats": Array [
                  Object {
                    "description": "Initialize a repo",
                    "line": "git init",
                  },
                ],
                "title": "Installation",
              },
              "onAddToFavorite": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SideNavPage
            errors={Object {}}
            onFormFieldChange={[Function]}
            onSignin={[Function]}
            onSignup={[Function]}
            user={
              Object {
                "firstName": "",
                "lastName": "",
                "password": "",
                "username": "",
              }
            }
            userAction={
              Object {
                "type": Symbol(INITIAL_USER_STATE),
              }
            }
          />,
          <SearchBar
            onSearchSheetCheat={[Function]}
          />,
          <MasonryComponent
            className="cheat-masonry-grid"
            disableImagesLoaded={false}
            elementType="div"
            enableResizableChildren={false}
            imagesLoadedOptions={Object {}}
            onLayoutComplete={[Function]}
            onRemoveComplete={[Function]}
            options={Object {}}
            updateOnEachImageLoad={false}
          >
            <Category
              category={
                Object {
                  "cheats": Array [
                    Object {
                      "description": "Initialize a repo",
                      "line": "git init",
                    },
                  ],
                  "title": "Installation",
                }
              }
              onAddToFavorite={[Function]}
            />
          </MasonryComponent>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "errors": Object {},
            "onFormFieldChange": [Function],
            "onSignin": [Function],
            "onSignup": [Function],
            "user": Object {
              "firstName": "",
              "lastName": "",
              "password": "",
              "username": "",
            },
            "userAction": Object {
              "type": Symbol(INITIAL_USER_STATE),
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onSearchSheetCheat": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Mon Jan 21 2019 13:38:26 GMT+0100 (West Africa Standard Time)",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Category
                category={
                  Object {
                    "cheats": Array [
                      Object {
                        "description": "Initialize a repo",
                        "line": "git init",
                      },
                    ],
                    "title": "Installation",
                  }
                }
                onAddToFavorite={[Function]}
              />,
            ],
            "className": "cheat-masonry-grid",
            "disableImagesLoaded": false,
            "elementType": "div",
            "enableResizableChildren": false,
            "imagesLoadedOptions": Object {},
            "onLayoutComplete": [Function],
            "onRemoveComplete": [Function],
            "options": Object {},
            "updateOnEachImageLoad": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "category": Object {
                  "cheats": Array [
                    Object {
                      "description": "Initialize a repo",
                      "line": "git init",
                    },
                  ],
                  "title": "Installation",
                },
                "onAddToFavorite": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
